src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs

```diff
@@ -286,72 +282,69 @@ private void EnsureInitialized()
 
-                    var extensionTypes = extension
-                        .ExportedTypes
-                        .Where(t => t.GetTypeInfo().IsClass && !t.GetTypeInfo().IsAbstract);
-
                     foreach (var feature in extension.ExtensionInfo.Features)
                     {
-                        var featureTypes = new HashSet<Type>();
-
-                        // Search for all types from the extensions that are not assigned to a different
-                        // feature.
-                        foreach (var type in extensionTypes)
+                        // Features can have no types
+                        if (typesByFeature.TryGetValue(feature.Id, out var featureTypes))
                         {
-                            string sourceFeature = GetSourceFeatureNameForType(type, extension.ExtensionInfo.Id);
-
-                            if (sourceFeature == feature.Id)
+                            foreach (var type in featureTypes)
                             {
-                                featureTypes.Add(type);
                                 _typeFeatureProvider.TryAdd(type, feature);
                             }
                         }
-
-                        // Search in other extensions for types that are assigned to this feature.
-                        var otherExtensionInfos = extensions.Where(x => x.Id != extension.ExtensionInfo.Id);
-
-                        foreach (var otherExtensionInfo in otherExtensionInfos)
+                        else
                         {
-                            var otherExtension = loadedExtensions[otherExtensionInfo.Id];
-                            foreach (var type in otherExtension.ExportedTypes)
-                            {
-                                string sourceFeature = GetSourceFeatureNameForType(type, null);
-
-                                if (sourceFeature == feature.Id)
-                                {
-                                    featureTypes.Add(type);
-                                    _typeFeatureProvider.TryAdd(type, feature);
-                                }
-                            }
+                            featureTypes = Array.Empty<Type>();
                         }
 
                         loadedFeatures.Add(feature.Id, new CompiledFeatureEntry(feature, featureTypes));
                     }
                 };
 
                 _extensions = loadedExtensions;
+
                 // Could we get rid of _allOrderedFeatureInfos and just have _features?
                 _features = loadedFeatures;
                 _allOrderedFeatureInfos = Order(loadedFeatures.Values.Select(x => x.FeatureInfo));
                 _isInitialized = true;
             }
         }
 
+        private bool IsComponentType(Type type)
+        {
+            var typeInfo = type.GetTypeInfo();
+            return typeInfo.IsClass && !typeInfo.IsAbstract && typeInfo.IsPublic;
```

Comment:
Just a first comment, normally, because we use assembly `ExportedTypes` which are all public, we don't need to check `IsPublic`. But it is as i remember, so not sure. But maybe safer to check it here, e.g if tomorrow we use `DefinedTypes`.